DTLSv1.3 support

rizlik opened the pull request:
# Description

DTLSv1.3 support.
[RFC 9147](https://datatracker.ietf.org/doc/rfc9147/)

# Known Issues
- No ConnectionID support
- It's vulnerable to some Denial of Service attacks
- Don't check the maximum number of failed decryption 
- Don't check if we reach the rekeying/encryption limit
- Client does not support v1.2/v1.3 dual stack

# Building

To build with DTLSv1.3 you can use `./configure --enable-dtls13 --enable-dtls`

# Testing

You can use `./example/client -v4 -u` and `./example/server -v4 -u` to use DTLSv1.3 with the features supported by the server and the client example programs. 

fe2154a5813c4d533550d1238b97e005ec9b4f61 adds test configurations that test the most part of the features. 

To simulate packet loss and re-ordering you can use the [udp proxy](https://github.com/wolfSSL/udp-proxy) tool. `scripts/dtls13.test` use this tool to perform various tests involving dropping and re-ordering of packets. 

# Checklist

 - [x] added tests
 - [ ] updated/added doxygen
 - [ ] updated appropriate READMEs
 - [ ] Updated manual and documentation

SparkiDev commented on src/dtls13.c:
```diff
@@ -0,0 +1,1000 @@
+
+    return method;
+}
+static int dtls13_rl_add_plaintext_header(WOLFSSL *ssl,
```
Formatting
If it's going to squeeze parameters over to the right like that, instead start the next lines with 4 spaces and then the parameters.

SparkiDev commented on src/dtls13.c:
```diff
@@ -0,0 +1,1000 @@
+    int ret;
+
+    ret = CheckAvailableSize(ssl, length);
+    if(ret != 0)
```
Formatting
Always a space after key words like if and while.

SparkiDev commented on src/dtls13.c:
```diff
@@ -0,0 +1,1000 @@
+
+        msg = output + DTLS_RECORD_HEADER_SZ;
+
+    } else {
```
Formatting
} else
}

SparkiDev commented on src/dtls13.c:
```diff
@@ -0,0 +1,1000 @@
+            return BUFFER_ERROR;
+
+        /* FIXME: ignore async ok flag for now */
+        send_length = BuildTls13Message(ssl, output,
```
Formatting
Similar to func decls, if it is going to squeeze right then put parameters on next line 4 more spaces in then original line.

SparkiDev commented on src/internal.c:
```diff
@@ -9356,11 +9389,123 @@ int CheckAvailableSize(WOLFSSL *ssl, int size)
+                                 RecordLayerHeader* rh, word16 *size)
+{
+    byte epoch_bits, seq_lo, seq_hi;
+    word16 header_lenght;
```
header_lenght -> header_length

SparkiDev commented on src/tls13.c:
```diff
@@ -7083,8 +7469,22 @@ static int SendTls13Finished(WOLFSSL* ssl)
+        if (dtlsRet != 0 && dtlsRet != WANT_WRITE)
+            return ret;
+
+        goto derive_key;
```
Don't use goto like this.
#ifdef WOLFSSL_DTLS13
    if (ssl->options.dtls) {
    }
    else
#endif
    {
    }

SparkiDev commented on wolfssl/internal.h:
```diff
@@ -5084,6 +5123,46 @@ WOLFSSL_LOCAL int oid2nid(word32 oid, int grp);
+WOLFSSL_LOCAL int dtls13_encrypt_record_number(WOLFSSL *ssl, byte *hdr);
+WOLFSSL_LOCAL int dtls13_is_unified_header(byte header_flags);
+WOLFSSL_LOCAL int dtls13_parse_unifed_record_layer(WOLFSSL *ssl,
+                                                   const byte *input,
```
This makes me think you have too many parameters.

rizlik replied:
yes I agree, I'll make a struct to hold the parameters

julek-wolfssl commented on src/dtls13.c:
```diff
@@ -0,0 +1,996 @@
+    if (ret != 0)
+        return ret;
+
+    for (i = 0; i < seq_length; ++i)
+        seq[i] ^= mask[i];
```
Let's use `xorbuf`.

julek-wolfssl commented on src/dtls13.c:
```diff
@@ -0,0 +1,996 @@
+    key = (side == WOLFSSL_SERVER_END) ?
+        ssl->keys.server_sn_key : ssl->keys.client_sn_key;
+
+    ret = wc_AesSetKey(&aes, key, ssl->specs.key_size, NULL, AES_ENCRYPTION);
+    if (ret != 0)
+        goto exit;
+
+    ret = wc_AesEncryptDirect(&aes, mask, ciphertext);
+    if (ret != 0)
+        goto exit;
+
+ exit:
+    wc_AesFree(&aes);
+    return ret;
```
In general we avoid using `goto`. You can use the following pattern here:
```
if (ret == 0)
    ret = wc_AesSetKey
if (ret == 0)
    ret = wc_AesEncryptDirect
```

julek-wolfssl commented on src/dtls13.c:
```diff
@@ -0,0 +1,996 @@
+    if (ssl->specs.bulk_cipher_algorithm != wolfssl_aes_gcm)
+        return WOLFSSL_NOT_IMPLEMENTED;
+
+    ret = wc_AesInit(&aes, ssl->heap, -1);
```
Use `INVALID_DEVID` here.

rizlik replied:
In the next PR the Aes object will be included in a dedicated Cipher in the SSL object.

julek-wolfssl commented on src/dtls13.c:
```diff
@@ -0,0 +1,996 @@
+    return 0;
+}
+
+/* this is a clone of DtlsMsgDrain */
```
Any reason they can't be consolidated into one implementation?

rizlik replied:
Not at the moment, I'll merge them if I can

julek-wolfssl commented on src/dtls13.c:
```diff
@@ -0,0 +1,996 @@
+        /* TODO: update the retransmission logic with new message processed */
+
+        /* FIXME: move in a better place. handle overflow? */
+        ssl->keys.dtls_expected_peer_handshake_number++;
```
Use `DtlsSEQIncrement`.

julek-wolfssl commented on src/dtls13.c:
```diff
@@ -0,0 +1,996 @@
+
+static int dtls13_next_message_complete(WOLFSSL *ssl) {
+    return ssl->dtls_rx_msg_list != NULL &&
+        ssl->dtls_rx_msg_list->fragSz >= ssl->dtls_rx_msg_list->sz;
```
Wouldn't `ssl->dtls_rx_msg_list->fragSz > ssl->dtls_rx_msg_list->sz` be an error? Is there a case where this is possible?

julek-wolfssl commented on src/dtls13.c:
```diff
@@ -0,0 +1,996 @@
+    if (ret != 0)
+        return ret;
+
+    buf = ssl->buffers.outputBuffer.buffer +
+        ssl->buffers.outputBuffer.length;
```
Would it make sense to add a `frag_is_in_output_buffer` check here as well? To make sure that in the future you don't accidentally pass in `outputBuffer` in `output`.

rizlik replied:
I think this will be invoked *only* after
```
    if (!frag_is_in_output_buffer(ssl, output))
        return dtls13_send_frag_from_buffer(ssl, output, length);
 ```
 this check. So it should be fine.

julek-wolfssl reviewed (changes requested):
Nice work! So far went through 1/3 of `dts13.c`. Will continue to review the rest and hopefully also add some more high level comments.

julek-wolfssl reviewed (changes requested):
Finished going through `dtls13.c`. I'll go through the rest in the next review.

julek-wolfssl commented on src/internal.c:
```diff
@@ -9356,11 +9389,125 @@ int CheckAvailableSize(WOLFSSL *ssl, int size)
+        return SEQUENCE_ERROR;
+    }
+
+    ssl->keys.curSeq_hi = ssl->keys.peer_sequence_number_hi;
+
+    c32toa(ssl->keys.peer_sequence_number_lo, new_seq_lo);
+    new_seq_lo[3] = seq_lo;
+    if (seq_present)
+        new_seq_lo[2] = seq_hi;
+    ato32(new_seq_lo, &ssl->keys.curSeq_lo);
```
Need to check if current epoch. Otherwise use `dtls_prev_sequence_number_hi` and `dtls_prev_sequence_number_lo`.

rizlik replied:
seq number will take another approach in future commits

julek-wolfssl commented on src/dtls13.c:
```diff
@@ -0,0 +1,996 @@
+        *seq_lo = seq_num[1];
+    } else {
+        *seq_hi_present = 0;
+        *seq_hi = seq_num[0];
```
Shouldn't this be setting `seq_lo` instead?

rizlik replied:
nice catch, thanks

julek-wolfssl commented on src/internal.c:
```diff
@@ -9356,11 +9389,125 @@ int CheckAvailableSize(WOLFSSL *ssl, int size)
+    ssl->keys.curSeq_hi = ssl->keys.peer_sequence_number_hi;
+
+    c32toa(ssl->keys.peer_sequence_number_lo, new_seq_lo);
+    new_seq_lo[3] = seq_lo;
```
Not always set in `dtls13_parse_unifed_record_layer` currently. Possible bug.

rizlik replied:
yes thanks

julek-wolfssl commented on src/tls13.c:
```diff
@@ -4641,25 +4821,36 @@ int DoTls13ClientHello(WOLFSSL* ssl, const byte* input, word32* inOutIdx,
-        ssl->version.minor = args->pv.minor;
+#endif /* WOLFSSL_DTLS13 */
+
+    if (!dtlsLegacyVersionChecked) {
+        if (args->pv.major > SSLv3_MAJOR || (args->pv.major == SSLv3_MAJOR &&
+                                             args->pv.minor >= TLSv1_3_MINOR)) {
+            args->pv.major = SSLv3_MAJOR;
+            args->pv.minor = TLSv1_2_MINOR;
+            wantDowngrade = 1;
+            ssl->version.minor = args->pv.minor;
+        }
+        /* Legacy version must be [ SSLv3_MAJOR, TLSv1_2_MINOR ] for TLS v1.3 */
+        else if (args->pv.major == SSLv3_MAJOR &&
+                 args->pv.minor < TLSv1_2_MINOR) {
+            wantDowngrade = 1;
+            ssl->version.minor = args->pv.minor;
+        }
```
Any reason for this to be outside the `WOLFSSL_DTLS13` area?

rizlik replied:
The goal was to reduce the amount of `ifdef`

julek-wolfssl replied:
Include this in the `WOLFSSL_DTLS13` section just above it.

rizlik replied:
This part was is needed by tls13. It was used the check to avoid #ifdef #else #endif macro

julek-wolfssl replied:
Ok now I see. Please re-instate the following comment before the inner `if` `/* Legacy protocol version cannot negotiate TLS 1.3 or higher. */`

julek-wolfssl commented on src/tls13.c:
```diff
@@ -5047,6 +5249,14 @@ int SendTls13ServerHello(WOLFSSL* ssl, byte extMsgType)
+                                     server_hello, 0);
+#endif /* WOLFSSL_DTLS13 */
+
+    ssl->buffers.outputBuffer.length += sendSz;
```
When `dtls13_handshake_send` is called this doesn't get advanced. Is there a reason I'm not seeing?

rizlik replied:
it will be advanced by `dtls13_send_one_fragment`

julek-wolfssl commented on src/tls13.c:
```diff
@@ -5070,14 +5280,23 @@ static int SendTls13EncryptedExtensions(WOLFSSL* ssl)
+#ifdef WOLFSSL_DTLS13
+    if (ssl->options.dtls) {
+        idx = dtls13_get_headers_length(ssl, encrypted_extensions);
+    } else
```
```
}
else
```

julek-wolfssl commented on src/tls13.c:
```diff
@@ -7006,6 +7345,33 @@ int DoTls13Finished(WOLFSSL* ssl, const byte* input, word32* inOutIdx,
+            }
+#endif
+
+            peerSeq->nextEpoch++;
```
This should be incremented by 2 (unless you have already implemented early data support).

julek-wolfssl commented on src/tls13.c:
```diff
@@ -5147,11 +5405,10 @@ static int SendTls13EncryptedExtensions(WOLFSSL* ssl)
 
     ssl->buffers.outputBuffer.length += sendSz;
 
-    ssl->options.serverState = SERVER_ENCRYPTED_EXTENSIONS_COMPLETE;
```
No state change for TLS 1.3?

rizlik replied:
good catch thanks!

julek-wolfssl commented on src/dtls13.c:
```diff
@@ -0,0 +1,996 @@
+ * dtls13_fragments_continue() - keep sending pending fragments
+ * @ssl: ssl object
+ */
+int dtls13_fragments_continue(WOLFSSL *ssl)
+{
+    int ret;
+
+    ret = dtls13_send_fragmented_internal(ssl);
+    if (ret != 0)
+        ssl->error = ret;
+
+    if (ret == 0)
+        ssl->keys.dtls_handshake_number++;
+
+    return ret;
+}
```
IMO it would be more elegant to have `dtls13_fragments_continue` return the error and the higher levels should set the `ssl->error`.

rizlik replied:
I agree thanks

julek-wolfssl reviewed (changes requested):
Final round of comments. Looks good. I don't have any high level comment for now except.

rizlik commented on src/dtls13.c:
```diff
@@ -0,0 +1,996 @@
+    is_protected = dtls13_type_is_encrypted(handshake_type);
+    record_header_length = dtls13_get_rl_header_length(ssl, is_protected);
+
+    if (length <= record_header_length)
+            return BUFFER_ERROR;
```
here length includes both headers and the actual payload size, so it should be at least recrd_header_length (don't know if we can arrive with a zero-length payload packet here). I don't know if the check is redundant or not, but it also makes sure that ```record_length = length - record_header_length;``` does not underflow

rizlik commented on src/dtls13.c:
```diff
@@ -0,0 +1,996 @@
+        frag_length = max_fragment - rl_header_length -
+            DTLS_HANDSHAKE_HEADER_SZ;
+
+        if (is_encrypted)
+            frag_length -= cipherExtraData(ssl);
```
oh nice, thanks.

rizlik commented on src/dtls13.c:
```diff
@@ -0,0 +1,996 @@
+        if (frag_length > remaining_size)
+            frag_length = remaining_size;
+
+        ret = CheckAvailableSize(ssl, max_fragment);
```
I think that `max_fragment` is equal to `rl_header_length + frag_length +  DTLS_HANDSHAKE_HEADER_SZ`,    because `frag_length` is calculated as: `frag_length = max_fragment -  rl_header_length - DTLS_HANDSHAKE_HEADER_SZ`. Of course if `max_fragment` takes in account cipher data, this must be changed to use all the available space.

rizlik commented on src/dtls13.c:
```diff
@@ -0,0 +1,996 @@
+    is_encrypted = dtls13_type_is_encrypted(handshake_type);
+    rl_header_length = dtls13_get_rl_header_length(ssl, is_encrypted);
+
+    if (length < rl_header_length)
+        return INCOMPLETE_DATA;
```
here `length` is the `length` of the message with headers included and make the ` ret = HashRaw(ssl, message + rl_header_length,
                      length - rl_header_length);` safe

rizlik commented on src/dtls13.c:
```diff
@@ -0,0 +1,996 @@
+ *
+ * returns the length of the record layer header in bytes.
+ */
+int dtls13_get_rl_header_length(WOLFSSL *ssl, int is_encrypted)
```
Yeah, but it will probably be used if we want to control with `ssl->options` how to build the variable header.

rizlik commented on src/dtls13.c:
```diff
@@ -0,0 +1,996 @@
+ * dtls13_get_headers_length() - return length of record + handshake header
+ * @type: type of handshake in the message
+ */
+int dtls13_get_headers_length(WOLFSSL *ssl, enum HandShakeType type)
```
Yeah, but it will probably be used if we want to control with ssl->options how to build the variable header.

rizlik replied:
removed the arg altogether, if needed we re-add it later.

rizlik commented on src/dtls13.c:
```diff
@@ -0,0 +1,996 @@
+            return LENGTH_ERROR;
+
+    seq_num = (byte*)(input + HEADER_FLAGS_SIZE);
+    /* cast is needed to override the const qualifier */
```
`ciphertext` cast can be removed, thanks.
Yes, it makes sense to change the parameter here and move the cast on the upper functions. The source of the `const` is `GetRecordHeader()`

julek-wolfssl commented:
I need to compile with `./configure --enable-dtls13 --enable-dtls` because otherwise I get build errors without the `--enable-dtls`.

julek-wolfssl commented on src/dtls13.c:
```diff
@@ -0,0 +1,1968 @@
+ *
+ * Further info rfc draft 43 sec 4.2.3
+ */
+int Dtls13EncryptRecordNumber(WOLFSSL *ssl, byte *hdr)
```
Somewhere before `Dtls13EncryptRecordNumber` there should be a check that the record length is at least an AES block size. Probably in `BuildTls13Message` but it would be better to pass in the `hdr` length and perform the check here.

rizlik replied:
I added the check here, but somewhere in BuildTls13Message we should implement the possibility to pad the record to satisfy the constrain

julek-wolfssl replied:
I agree but that is a future TODO :sunglasses:

rizlik replied:
Hi @julek-wolfssl , this check is added. Do you consider this resolved?

julek-wolfssl commented on src/dtls13.c:
```diff
@@ -0,0 +1,1968 @@
+    hdr->unifiedHdrFlags |= L_BIT;
+
+    seqNumber = ssl->dtls13EncryptEpoch->nextSeqNumberLo;
+    hdr->sequenceNumber[0] = (seqNumber >> 16) & 0xff;
+    hdr->sequenceNumber[1] = seqNumber & 0xff;
```
I'm fairly certain I made a mistake here since it should be `>> 8`. That's why its best to use `c16toa`.

julek-wolfssl commented on src/dtls13.c:
```diff
@@ -0,0 +1,996 @@
+        ssl->keys.dtls_expected_peer_handshake_number)
+        return WOLFSSL_NOT_IMPLEMENTED;
+
+    /* FIXME: client_hello should be handled differently */
```
Have you tested your implementation against a scenario where you receive a `ClientHello` -> we reply with `HelloRetryRequest` and then the client sends a new `ClientHello` without the cookie? Receiving bogus `ClientHello` should not brick a `WOLFSSL` object.

rizlik replied:
HelloRetryRequest isn't supported yet.

julek-wolfssl commented:
Please provide instruction of how to best test the re-transmission mechanisms.

julek-wolfssl commented on src/dtls13.c:
```diff
@@ -0,0 +1,1968 @@
+}
+
+static int Dtls13RtxAddAck(
+    WOLFSSL *ssl, Dtls13RtxFSM *fsm, word32 recordNumber[2])
```
Be careful with array arguments. This doesn't actually mean anything other than that this is a pointer. Better to have `word32* recordNumber` with documentation. Same goes for all other array arguments.

rizlik replied:
yeah, I wonder if record numbers deserve their own type. Epoch size is likely going to grow in the final RFC and we may want to save only the last bits of it.

julek-wolfssl replied:
>record numbers deserve their own type

This is a good idea. Let's do that.

julek-wolfssl commented on src/dtls13.c:
```diff
@@ -0,0 +1,1968 @@
+    WOLFSSL_ENTER("Dtls13RtxRecordRecvd");
+
+    ret = Dtls13RtxAddAck(ssl, fsm, recordNumber);
+    if (ret != 0) {
+        WOLFSSL_MSG("can't save ack fragment");
+    }
```
Not erroring out here could potentially cause deadlocks where we won't send the require ACKs. The peer will keep sending its last flight and will not receive an ACK for it ever.

julek-wolfssl commented on wolfssl/internal.h:
```diff
@@ -4712,7 +4836,10 @@ enum ContentType {
-    application_data   = 23
+    application_data   = 23,
+#ifdef WOLFSSL_DTLS13
+    ack                = 25,
```
This should be 26. It is only refered to as 25 in
https://www.ietf.org/archive/id/draft-ietf-tls-dtls13-43.html#ack-msg
But in the following it is proposed as 26
https://www.ietf.org/archive/id/draft-ietf-tls-dtls13-43.html#name-demultiplexing-dtls-records
https://www.ietf.org/archive/id/draft-ietf-tls-dtls13-43.html#name-iana-considerations
This is an error in the draft rfc by the way. You should report it to them.

julek-wolfssl replied:
Ok. So its settled that this will have value 26.

julek-wolfssl commented on src/dtls13.c:
```diff
@@ -0,0 +1,1968 @@
+    }
+}
+
+static int Dtls13RtxIsLastFligh(WOLFSSL *ssl)
```
Typo

rizlik replied:
function removed in next iteration

julek-wolfssl commented on src/dtls13.c:
```diff
@@ -0,0 +1,1968 @@
+    return 0;
+}
+
+static int Dtls13RtxIsTrackedByRn(Dtls13RtxRecord *r, const word32 seq[2])
```
Let's make `r` a `const Dtls13RtxRecord*`.

julek-wolfssl commented on src/dtls13.c:
```diff
@@ -0,0 +1,1968 @@
+    if (head == NULL)
+        return 0;
+
+    /* check head first */
+    if (Dtls13RtxIsTrackedByRn(head, seq)) {
+        fsm->rtxRecords = head->next;
+        Dtls13FreeRtxBufferRecord(ssl, head);
+        return 0;
+    }
+
+    while (head->next != NULL) {
+        r = head->next;
+        if (Dtls13RtxIsTrackedByRn(r, seq)) {
+            head->next = r->next;
+            Dtls13FreeRtxBufferRecord(ssl, r);
+            return 0;
+        }
+
+        head = head->next;
+    }
```
Let's make this one `while` loop
```
while (head != NULL) {
  ...
  head = head->next;
}
```

rizlik replied:
If I do, I should keep track of the preceding element to  do the unlink `head->next = r->next;`

julek-wolfssl commented on wolfssl/internal.h:
```diff
@@ -4241,8 +4263,90 @@ typedef enum EarlyDataState {
+    byte isValid;
+} Dtls13Epoch;
+
+#define DTLS13_EPOCH_SIZE 4
+
+#define DTLS13_RETRANS_RN_SIZE 3
```
Use this template so that this can be customized with compile time flags
```
#ifndef DTLS13_RETRANS_RN_SIZE
#define DTLS13_RETRANS_RN_SIZE 3
#endif
```

julek-wolfssl commented on src/dtls13.c:
```diff
@@ -0,0 +1,1968 @@
+    byte length[2];
+} Dtls13RecordCiphertextHeader;
+
+#define MAX_SEQ_SIZE 2
+#define LEN_SIZE 2
+#define RN_MASK_SIZE 16
+#define HEADER_FLAGS_SIZE 1
+#define SEQ_16_LEN 2
+#define SEQ_8_LEN 1
+#define SEQ_FULL_SIZE 4
```
Add a `DTLS13_` prefix because they look very close to asn defines ex: `MAX_SEQ_SZ` or `MAX_LENGTH_SZ`.

julek-wolfssl commented on src/dtls13.c:
```diff
@@ -0,0 +1,1968 @@
+#define LEN_SIZE 2
+#define RN_MASK_SIZE 16
+#define HEADER_FLAGS_SIZE 1
+#define SEQ_16_LEN 2
+#define SEQ_8_LEN 1
```
Use `OPAQUE8_LEN` and `OPAQUE16_LEN` instead maybe?

rizlik replied:
I'll use these specific ones, as opaque as a specific meaning in TLS type description language.

rizlik commented on wolfssl/internal.h:
```diff
@@ -4241,8 +4263,90 @@ typedef enum EarlyDataState {
+typedef struct Dtls13RtxFSM {
+    enum Dtls13RtxFsmState state;
+    Dtls13RtxRecord *rtxRecords;
+    Dtls13RecordNumber *ackRecords;
```
That is a good idea. I'll try to refactor that soon

rizlik replied:
In the end, I opted for using a dynamic allocation as the record number became bigger in the final draft

rizlik commented:
- Added support for post-handshake messages (KeyUpdate, Certificate, NewSessionTickets)
- Addressed most of the feedbacks

julek-wolfssl commented on src/dtls13.c:
```diff
@@ -0,0 +1,2187 @@
+        if (c->chacha == NULL)
+            return BAD_STATE_E;
+
+        /* assuming CIPHER[0..3] should be interpreted as big endian 32-bits
+           integer. The draft rfc isn't really clear on that. See sec 4.2.3 */
+        ato32(ciphertext, &counter);
+        ret = wc_Chacha_SetIV(c->chacha, &ciphertext[4], counter);
+        if (ret != 0)
+            return ret;
```
I know we discussed this already but I will leave this comment to remember about this. I think that the byte stream should be interpreted as a little endian number per [Section 2.3 of the Chacha RFC](https://datatracker.ietf.org/doc/html/rfc7539#section-2.3).

julek-wolfssl replied:
I still don't agree with this.

julek-wolfssl replied:
Resolved to be LE.

rizlik commented on src/internal.c:
```diff
@@ -20148,12 +20606,30 @@ int SendAlert(WOLFSSL* ssl, int severity, int type)
      */
     if (IsEncryptionOn(ssl, 1)) {
+#ifdef WOLFSSL_DTLS13
+        if (ssl->dtls13EncryptEpoch->epochNumber != ssl->dtls13Epoch) {
+            ret = Dtls13SetEpochKeys(ssl, ssl->dtls13Epoch, ENCRYPT_SIDE_ONLY);
+            if (ret != 0)
+                return ret;
+        }
```
definitely, thanks

rizlik commented on src/internal.c:
```diff
@@ -2608,6 +2629,30 @@ void FreeCiphers(WOLFSSL* ssl)
+    }
+#endif /* BUILD_AES */
+#ifdef HAVE_CHACHA
+        XFREE(ssl->dtlsRecordNumberEncrypt.chacha,
+              ssl->heap, DYNAMIC_TYPE_CIPHER);
+        XFREE(ssl->dtlsRecordNumberDecrypt.chacha,
+              ssl->heap, DYNAMIC_TYPE_CIPHER);
```
I followed the pattern of some lines above here. I supposed freeing NULL pointers should not be a problem.

julek-wolfssl replied:
`free(NULL)` is a no-op according to the `man` page but I'm worried if it is safe on embedded systems.

julek-wolfssl commented on src/dtls13.c:
```diff
@@ -0,0 +1,2186 @@
+    (void)hs;
+
+#ifndef NO_WOLFSSL_SERVER
+    if (ssl->options.side == WOLFSSL_SERVER_END && (hs == finished))
```
I'm still not convinced that only `finished` should be marked as needing an explicit ack. This combined with the following section
```
    if (ssl->keys.dtls_peer_handshake_number ==
            ssl->keys.dtls_expected_peer_handshake_number &&
        Dtls13RtxNeedsExplicitAck(ssl, hs))
        fsm->sendAcks = 1;
```
feels like someone might break this in the future because the connection between these two sections is not obvious. I'll still ask you to consider marking all of the client's final flight as needing explicit acks and thinking of another way to not send them all immediately.

rizlik replied:
It may be the naming it's also unfortunate, `Dtls13RtxNeedsExplicitAck`  more than "does the message need an explicit ACK?" means "Is it the last message of a flight that needs an explicit ACK?". 

How do you see moving the `if(Dtls13RtxNeedsExplicitAck) sendAck = 1` check after processing the message? Do you think it would make more sense?

On top of my mind, the only point where we may need to send an ACK immediately for `Cert/CertVerify` is in case the server:
1. receives and buffers the finished message, it sends an ACK for the finished
2. client resends `cert/cert verify`
3. server processes everything, sends the ACKs after processing the finished message buffered  previously
4. The ACK on point 3 is lost
5. client resends `cert/cert verify`

At this point, we detect the `cert/cert verify` as a retransmission. In this case, when the `sendMoreAck` option is on, we ack those messages immediately, otherwise we will send the ack when (fast) timeout expires.

It looks sound to me. Am I missing something?

julek-wolfssl replied:
>It may be the naming it's also unfortunate, Dtls13RtxNeedsExplicitAck more than "does the message need an explicit ACK?" means "Is it the last message of a flight that needs an explicit ACK?".

In that case I would recommend changing it to "Dtls13IsLastMsgInFlight" or something so that its more obvious that that is the purpose of this check. Possibly split it into two functions (if needed) where one checks if it needs to be acked and the other if it is the last msg of a flight.

>How do you see moving the if(Dtls13RtxNeedsExplicitAck) sendAck = 1 check after processing the message? Do you think it would make more sense?

I think that this is a better way to go about doing this. Doing the check as soon as the msg arrives forces you to account for network issues which are making everything more complicated.

> [the rest]

I agree that we don't need to send acks immediately for other msgs. In fact we should, as you mentioned, only do it on a fast timeout. I feel like the naming is misleading and the connection with the section of code I mentioned isn't obvious.

rizlik replied:
Hi @julek-wolfssl ,
can you please re-check this with the last version?

julek-wolfssl replied:
Please just add a comment to this function that it expects that the entire last flight will be included in the `finished` ACK.

julek-wolfssl commented on src/wolfio.c:
```diff
@@ -330,23 +329,43 @@ int EmbedReceiveFrom(WOLFSSL *ssl, char *buf, int sz, void *ctx)
+    if (!doDtlsTimeout)
         dtls_timeout = 0;
 
     if (!wolfSSL_get_using_nonblock(ssl)) {
```
Please also consider implementing a timeout mechanism for non-blocking sockets. You can leverage the `else if(IsSCR(ssl))` section to accomplish this. It currently only uses a `LowResTimer` but its better than nothing.

rizlik replied:
I think that for non-blocking sockets, the application should take care of the dtls timeout. at least in this first beta version

julek-wolfssl commented on src/dtls13.c:
```diff
@@ -608,28 +600,49 @@ static int Dtls13DetectDisruption(WOLFSSL *ssl, word32 fragOffset)
     }
+}
+
+/* recordNumber is a size 2 array */
```
Comment outdated.

julek-wolfssl commented on src/dtls13.c:
```diff
@@ -608,28 +600,49 @@ static int Dtls13DetectDisruption(WOLFSSL *ssl, word32 fragOffset)
+}
+
+/* recordNumber is a size 2 array */
+static int Dtls13RtxRecordRecvd(
+    WOLFSSL *ssl, enum HandShakeType hs, Dtls13RtxFSM *fsm, word32 fragOffset)
```
Consider renaming this to `Dtls13MsgRecvd` because this doesn't have anything to do with the record layer anymore and it doesn't necessary have to be a rtx msg.

julek-wolfssl replied:
So it is now `Dtls13RtxMsgRecvd`. This is definitely pedantic but is the `Rtx` necessary? It is a generic function that doesn't only handle `Rtx`.

julek-wolfssl replied:
The `WOLFSSL_ENTER` is now outdated. Please update.

julek-wolfssl commented on src/internal.c:
```diff
@@ -17365,10 +17712,36 @@ int ProcessReplyEx(WOLFSSL* ssl, int allowSocketErr)
+
+                    continue;
+                }
+                Dtls13UpdateWindow(ssl);
```
Need to check return value.

julek-wolfssl commented on src/dtls13.c:
```diff
@@ -0,0 +1,2209 @@
+    return 0;
+}
+
+static int Dtls13RtxLastMsgExplicitFlight(WOLFSSL *ssl, enum HandShakeType hs)
```
I feel that this function name is even more cryptic than before. Something more like `Dtls13NeedsAck` with a comment explaining that this function checks after processing which handshake message we are obligated to send acks.

rizlik replied:
you're right. I'll handle the rename of this and other functions soon (I may do this while squashing the commits in a more meaningful way)

SparkiDev commented on examples/server/server.c:
```diff
@@ -2131,13 +2140,22 @@ THREAD_RETURN WOLFSSL_THREAD server_test(void* args)
+#ifdef WOLFSSL_DTLS13
+                version = -4;
+#else
+                err_sys_ex(runWithErrors, "Bad DTSL version");
```
DTSL -> DTLS

SparkiDev commented on src/dtls13.c:
```diff
@@ -0,0 +1,2432 @@
+{
+    (void)heap;
+
+    WOLFSSL_ENTER("DTSLv23_server_method_ex");
```
DTSL -> DTLS

rizlik replied:
removed the function altogether

SparkiDev commented on src/dtls13.c:
```diff
@@ -0,0 +1,2432 @@
+
+static int Dtls13TypeIsEncrypted(enum HandShakeType hs_type)
+{
+    switch (hs_type) {
```
int ret = 0;
And then one return from function.

SparkiDev commented on src/dtls13.c:
```diff
@@ -0,0 +1,2432 @@
+}
+
+static int Dtls13GetRnMask(
+    WOLFSSL *ssl, const byte *ciphertext, byte *mask, enum rnDirection dir)
```
Put argument definitions on line above until overflow.

rizlik replied:
Did just for this for now. Should I apply that to every function? (I was letting the `clangd-formatter` take the decision)

SparkiDev replied:
Unfortunately formatters don't follow our rules. But sometimes they have options that will work.
I would like to see all functions match our convention of filling up the line and continue on next.

If you find options that make the code look like we want, let us know!
It would be good to be able to run a formatter over all the code to fix the formatting everywhere!

SparkiDev commented on src/dtls13.c:
```diff
@@ -0,0 +1,2432 @@
+    /* we need to send ACKs on the last message of a flight that needs explicit
+       acknowledgment */
+    if (Dtls13RtxMsgNeedsAck(ssl, hs))
+        ssl->dtls13Rtx.sendAcks = 1;
```
ssl->dtl13Rrx.sendAcks = Dtls13RtxMsgNeedsAck(ssl, hs);

SparkiDev commented on src/dtls13.c:
```diff
@@ -0,0 +1,2432 @@
+    return 0;
+}
+
+static void Dtls13MsgProcessed(WOLFSSL *ssl, enum HandShakeType hs)
```
Maybe Dtls13MsgWasProcessed()?

SparkiDev commented on src/dtls13.c:
```diff
@@ -0,0 +1,2432 @@
+        return ret;
+
+    idx += OPAQUE32_LEN;
+
```
length -= OPAQUE32_LEN;
and below then passing only length in.
Makes it cleaner to read.

SparkiDev commented on src/dtls13.c:
```diff
@@ -0,0 +1,2432 @@
+
+static int Dtls13DetectDisruption(WOLFSSL *ssl, word32 fragOffset)
+{
+    /* retranmisssion. The other peer may have lost our flight or our ACKs. We
```
retranmisssion -> Retransmission
'We not account' -> 'We don't account'?

SparkiDev commented on src/dtls13.c:
```diff
@@ -0,0 +1,2432 @@
+    recordHeaderLength = Dtls13GetRlHeaderLength(isProtected);
+
+    if (length <= recordHeaderLength)
+            return BUFFER_ERROR;
```
spaces

SparkiDev commented on src/dtls13.c:
```diff
@@ -0,0 +1,2432 @@
+}
+
+int Dtls13RlAddPlaintextHeader(
+    WOLFSSL *ssl, byte *out, enum ContentType content_type, size_t length)
```
WOLFSSL *ssl => WOLFSSL* ssl
consistency through the file.

SparkiDev commented on src/dtls13.c:
```diff
@@ -0,0 +1,2432 @@
+
+    remainingSize = ssl->dtls13MessageLength - ssl->dtls13FragOffset;
+
+    while(remainingSize > 0) {
```
space after while

rizlik replied:
missed that! I'm gonna fix it in the next review cycle

SparkiDev commented on src/dtls13.c:
```diff
@@ -0,0 +1,2432 @@
+    return ret;
+}
+
+#define DTLS13_MIN_RTX_INTERVAL 1
```
Move to top.

rizlik commented on src/dtls13.c:
```diff
@@ -0,0 +1,2432 @@
+
+static inline void Dtls13ToBigEndian(word64 in, byte *out)
+{
+    word64 *_out = (word64*)out;
```
Do we have any easy way to force the alignment of a variable? Using a variable on the stack guarantees that the variable is aligned?

julek-wolfssl replied:
You can't force alignment on an arbitrary `char` pointer (I mean you can but that means you have to advance the pointer to the next alignment boundary and you can't do that here). I had a similar issue with session tickets but the other direction where I wanted to cast the input to a bigger struct than `char` but either the compiler didn't like that or it segfaults. The only solution is to write out the `word64` byte by byte.

rizlik replied:
I didn't fix this yet.

rizlik commented on src/internal.c:
```diff
@@ -29594,7 +30186,18 @@ static int DoSessionTicket(WOLFSSL* ssl, const byte* input, word32* inOutIdx,
+
+#ifdef WOLFSSL_DTLS13
+            /* DTLS versions increase backwards (-1,-2,-3) ecc  */
+            if (ssl->options.dtls)
```
Don't understand this, I don't see any else branch.

SparkiDev replied:
Don't know what I was thinking here.

rizlik commented on src/dtls13.c:
```diff
@@ -0,0 +1,2432 @@
+    return method;
+}
+
+WOLFSSL_METHOD *wolfDTLSv23_server_method_ex(void *heap)
```
oh, thanks for pointing it out. I did change it. The client-side is still missing for now

rizlik commented on src/dtls13.c:
```diff
@@ -0,0 +1,2432 @@
+    return method;
+}
+
+static inline void Dtls13ToBigEndian(word64 in, byte *out)
```
I introduced a wrapper for word64 types and refactor the code that uses word64. here is the commit that introduces the wrapper : e82526ebf7bc72700c146be7de6307d7d80f76ca

rizlik commented on src/internal.c:
```diff
@@ -9357,11 +9467,182 @@ int CheckAvailableSize(WOLFSSL *ssl, int size)
+    /* DTLSv1.3 PlainText records use DTLSv1.2 sequence number encoding. Update
+       the DTLv1.3 word64 version as well */
+    ssl->keys.curSeq =
+        ((word64)(ssl->keys.curSeq_hi) << 32) | ssl->keys.curSeq_lo;
```
refactored to use `w64wrapper`

rizlik commented:
I did address @SparkiDev comments. 
I added a `w64wrapper` that can be used to abstract 64bits simple integer operations and that work on machines that haven't 64bit native support. It's introduced in the commit dac63e8806b7d677eef508728aea870a24b2a9e6. All the other changes are in the commits that start with "fixup!" so it's easier to check them. Let me know if you prefer to squash them in their respective previous commit.

julek-wolfssl commented on src/wolfio.c:
```diff
@@ -381,23 +381,48 @@ int EmbedReceiveFrom(WOLFSSL *ssl, char *buf, int sz, void *ctx)
             XMEMSET(&timeout, 0, sizeof(timeout));
-            timeout.tv_sec = dtls_timeout;
+#ifdef WOLFSSL_DTLS13
+            if (wolfSSL_dtls_13_use_quick_timeout(ssl) &&
+                IsAtLeastTLSv1_3(ssl->version)) {
+                if (dtls_timeout >= 4)
+                    timeout.tv_sec = dtls_timeout / 4;
+                else
+                    timeout.tv_usec = dtls_timeout * 1000000 / 4;
+            }
```
An idea I had just now is that it would be good to document this section for users. What the quick timeout means and that the `4` is an arbitrary value that users should define according to their own needs.

rizlik replied:
Good idea, I will add that together to other documentation

rizlik commented:
> @rizlik , consider doing a rebase to resolve the merge conflicts. It also might be a good time to squash these commits?

rebase and squashing done. I addressed some issues with version negotiation support as well.

SparkiDev commented on src/dtls13.c:
```diff
@@ -0,0 +1,2442 @@
+    return ret;
+}
+
+static int Dtls13InitAesCipher(WOLFSSL* ssl, Ciphers *cipher)
```
Have key passed in too.

SparkiDev commented on src/dtls13.c:
```diff
@@ -0,0 +1,2442 @@
+
+int Dtls13SetRecordNumberKeys(WOLFSSL *ssl, enum encrypt_side side)
+{
+    Ciphers *enc = NULL;
```
Using Ciphers seems a bit excessive.
RecNumCiphers {
    Aes* aes;
    ChaCha* chacha;
};

or

RecNumCiphers {
    union {
        Aes aes;
        ChaCha chacha;
   } alg;
};

SparkiDev commented on src/dtls13.c:
```diff
@@ -0,0 +1,2442 @@
+            if (ret != 0)
+                return ret;
+
+            if (ssl->options.side == WOLFSSL_CLIENT_END)
```
Can be done outside AES and Chacha specific code.

SparkiDev commented on src/dtls13.c:
```diff
@@ -0,0 +1,2442 @@
+                key = ssl->keys.server_sn_key;
+
+#ifdef WOLFSSL_DEBUG_TLS
+            WOLFSSL_MSG("Provisioning Record Number enc key:");
```
Say AES or ChaCha.

SparkiDev commented on src/dtls13.c:
```diff
@@ -0,0 +1,2442 @@
+ *
+ * returns the length of the record layer header in bytes.
+ */
+int Dtls13GetRlHeaderLength(int isEncrypted)
```
Consider making this a macro.

SparkiDev replied:
OK. The macro would go into internal.c.

SparkiDev commented on src/dtls13.c:
```diff
@@ -0,0 +1,2442 @@
+    for (i = 0; i < length; i += DTLS13_RN_SIZE) {
+        ato64(ackMessage + i, &epoch);
+        ato64(ackMessage + i + OPAQUE64_LEN, &seq);
+        Dtls13RtxRemoveRecord(ssl, epoch, seq);
```
Return ignored so consider making function void return.

rizlik commented on src/dtls13.c:
```diff
@@ -0,0 +1,2442 @@
+            ssl->keys.dtls_handshake_number++;
+    }
+    else {
+        ret = Dtls13SendFragmented(
```
We can't as we need to hash, setup a copy of the buffer and save some state before calling Dtls13SendFragmentedInternal()

SparkiDev replied:
Yeah, confused function names.

rizlik commented on src/dtls13.c:
```diff
@@ -0,0 +1,2442 @@
+    return ret;
+}
+
+static void Dtls13RtxMoveToEndOfList(Dtls13RtxRecord **listPtr,
```
Thanks now it looks way simpler and cleaner. In the future, If we use list structure in other parts of the code it may be helpful to abstract the data structure and share the code across the codebase.

SparkiDev commented on src/internal.c:
```diff
@@ -2088,7 +2109,17 @@ int InitSSL_Ctx(WOLFSSL_CTX* ctx, WOLFSSL_METHOD* method, void* heap)
+    }
+    else
+#endif /* WOLFSSL_DTLS */
+        {
```
Pull back 4 spaces in line with if

SparkiDev commented on src/ssl.c:
```diff
@@ -1120,6 +1120,13 @@ int wolfSSL_mutual_auth(WOLFSSL* ssl, int req)
+WOLFSSL_API
+int wolfSSL_is_peer_auth_good(WOLFSSL* ssl)
+{
+    return !!ssl->options.peerAuthGood;
```
peerAuthGood is 1 bit of a word16.
Don't see the need for !!

SparkiDev commented on src/tls13.c:
```diff
@@ -3409,18 +3578,38 @@ int DoTls13ServerHello(WOLFSSL* ssl, const byte* input, word32* inOutIdx,
+#endif /* WOLFSSL_DTLS13 */
+
+        if (wantDowngrade && ssl->options.downgrade)
+            {
```
Formatting looks wrong and brace should be on line above.

SparkiDev commented on src/tls13.c:
```diff
@@ -5932,18 +6306,39 @@ static int SendTls13Certificate(WOLFSSL* ssl)
-            else
+#ifdef WOLFSSL_DTLS13
+            else if (ssl->options.dtls){
+                /* short-cirtuit the fragmentation logic here. DTLS
```
short_cirtuit => short_circuit

SparkiDev commented on src/tls13.c:
```diff
@@ -6032,6 +6427,14 @@ static int SendTls13Certificate(WOLFSSL* ssl)
+            ret = Dtls13HandshakeSend(ssl, output, sendSz, i, certificate, 1);
+        } else
+#endif /* WOLFSSL_DTLS13 */
+        {
```
Formatting: push block in by 4 spaces.
Also:
}
else

SparkiDev commented on src/tls13.c:
```diff
@@ -9735,6 +10484,12 @@ int wolfSSL_accept_TLSv13(WOLFSSL* ssl)
+
+#ifdef WOLFSSL_DTLS13
+            if (ssl->options.dtls)
+                ssl->dtls13SendingAckOrRtx =0;
```
Formatting: space after =

rizlik commented:
addressed last comments and fixed a small bug (implementation didn't send an empty ack if the ServerHello got lost after an HelloRetryRequest)

dgarske reviewed (changes requested):
Looks good. Please consider squashing this down to no more than 5 commits. Thanks

dgarske commented on IDE/WIN/wolfssl-fips.vcxproj:
```diff
@@ -307,6 +307,8 @@
     <ClCompile Include="..\..\wolfcrypt\src\signature.c" />
     <ClCompile Include="..\..\src\ssl.c" />
     <ClCompile Include="..\..\src\tls.c" />
+    <ClCompile Include="..\..\src\tls13.c" />
```
Yes it is okay. Anything outside the wolfCrypt FIPS boundary like TLS/DTLS can change and should be a supported use/test case.

dgarske reviewed (changes requested):
Looks good and tests out.

Just a few whitespace issues:

```
diff --git a/examples/client/client-filter-progress.sh b/examples/client/client-filter-progress.sh
index f6fcb3445..11bcdb191 100755
--- a/examples/client/client-filter-progress.sh
+++ b/examples/client/client-filter-progress.sh
@@ -1,7 +1,6 @@
 #!/bin/bash
-  
+
 SCRIPT=`realpath $0`
 SCRIPTPATH=`dirname $SCRIPT`
 
 $SCRIPTPATH/client "$@" 2> >(grep --color=auto -P 'processing |Entering SendTls13|Entering SendDtls13|got |Entering SSL_write|Entering SendAlert|Dtls13Rtx One Record')
-
diff --git a/src/internal.c b/src/internal.c
index 1b470b8ef..2e3cb79b5 100644
--- a/src/internal.c
+++ b/src/internal.c
@@ -9894,7 +9894,7 @@ int GetDtlsHandShakeHeader(WOLFSSL* ssl, const byte* input,
 
     if ((ssl->curRL.pvMajor != ssl->version.major) ||
         (!IsAtLeastTLSv1_3(ssl->version) && ssl->curRL.pvMinor != ssl->version.minor) ||
-        (IsAtLeastTLSv1_3(ssl->version) && ssl->curRL.pvMinor != DTLSv1_2_MINOR) 
+        (IsAtLeastTLSv1_3(ssl->version) && ssl->curRL.pvMinor != DTLSv1_2_MINOR)
         ) {
         if (*type != client_hello && *type != hello_verify_request) {
             WOLFSSL_ERROR(VERSION_ERROR);
@@ -18312,7 +18312,7 @@ int ProcessReplyEx(WOLFSSL* ssl, int allowSocketErr)
                 /* For TLS v1.1 the block size and explcit IV are added to idx,
                  * so it needs to be included in this limit check */
                 if (!IsAtLeastTLSv1_3(ssl->version)
-                        && ssl->curSize - ssl->keys.padSz - 
+                        && ssl->curSize - ssl->keys.padSz -
                             (ssl->buffers.inputBuffer.idx - startIdx)
                                 > MAX_PLAINTEXT_SZ
 #ifdef WOLFSSL_ASYNC_CRYPT
```

Also do you still want `./examples/client/client-filter-progress.sh` to be included? Perhaps the ./scripts dir is a better place for it?

rizlik commented:
> Just a few whitespace issues:

Fixed, Thanks!
> Also do you still want ./examples/client/client-filter-progress.sh to be included? Perhaps the ./scripts dir is a better place for it?

I removed those scripts.

Also other minor changes:
added minimal documentation, added a small fixup on quick timeout, and did some renaming for consistency.

julek-wolfssl commented on src/wolfio.c:
```diff
@@ -381,23 +381,48 @@ int EmbedReceiveFrom(WOLFSSL *ssl, char *buf, int sz, void *ctx)
     if (!wolfSSL_get_using_nonblock(ssl)) {
         #ifdef USE_WINDOWS_API
             DWORD timeout = dtls_timeout * 1000;
+#ifdef WOLFSSL_DTLS13
```
Please line up the indentation to match the outer pre-proc guards.

julek-wolfssl reviewed (changes requested):
Hi @rizlik, just a few comments from a quick pass through the current state of the PR.

dgarske reviewed (changes requested):
The examples report `SSL version is unknown`

```
./configure --enable-dtls13 --enable-dtls && make

./examples/server/server -v4 -u
SSL version is unknown
SSL cipher suite is TLS_AES_128_GCM_SHA256
SSL curve name is SECP256R1
Client message: hello wolfssl!

./examples/client/client -v4 -u
SSL version is unknown
SSL cipher suite is TLS_AES_128_GCM_SHA256
SSL curve name is SECP256R1
I hear you fa shizzle!
```

rizlik commented on examples/client/client.c:
```diff
@@ -4004,8 +4047,31 @@ THREAD_RETURN WOLFSSL_THREAD client_test(void* args)
+                }
+            }
+        }
+#endif /* WOLFSSL_DTLS13 */
```
Indeed that part of the code, besides busy looping, is broken on non-blocking sockets because it doesn't take care of DTLS timeouts. I fixed that in the last version of the code.

I still didn't implement a timeout, I think hanging tests can happen also on application data or current DTLS tests, so a more general approach is needed. I think a good solution may be to implement a general "test" timeout in `suites.c`.

julek-wolfssl replied:
Has this section been removed?

rizlik replied:
Yes,
because of the removal of the post-handshake authentication check (for the reason that the server doesn't have an easy way to check when post-handshake authentication succeeds), it isn't necessary anymore. It will be likely re-added in the future to enable checks that rely on the client to properly send pending messages before shutdown.

dgarske reviewed (changes requested):
```
[all-gcc-c99] [4 of 22] [62d31e68d9]
    configure...   real 0m16.422s  user 0m11.481s  sys 0m6.766s
    build...src/internal.c: In function ‘GetRecordHeader’:
src/internal.c:9711:5: error: ISO C90 forbids mixed declarations and code [-Werror=declaration-after-statement]
 9711 |     int ret;
      |     ^~~
src/tls13.c: In function ‘SendTls13Finished’:
src/tls13.c:7740:9: error: ISO C90 forbids mixed declarations and code [-Werror=declaration-after-statement]
 7740 |         byte storeTrafficDecKeys = ssl->earlyData == no_early_data;
      |         ^~~~
src/tls13.c: In function ‘wolfSSL_connect_TLSv13’:
src/tls13.c:9342:5: error: ISO C90 forbids mixed declarations and code [-Werror=declaration-after-statement]
 9342 |     int advanceState;
      |     ^~~
src/tls13.c: In function ‘wolfSSL_accept_TLSv13’:
src/tls13.c:10366:5: error: ISO C90 forbids mixed declarations and code [-Werror=declaration-after-statement]
10366 |     int advanceState;
      |     ^~~
cc1: all warnings being treated as errors
```

dgarske commented on src/tls13.c:
```diff
@@ -3409,18 +3578,37 @@ int DoTls13ServerHello(WOLFSSL* ssl, const byte* input, word32* inOutIdx,
+
     if (args->pv.major != ssl->version.major ||
-        args->pv.minor != TLSv1_2_MINOR) {
+        args->pv.minor != tls12minor) {
```
This is happening with async. `byte tls12minor;` should be moved to `struct Dsh13Args`
```
./configure --enable-all --enable-asynccrypt && make

    build...src/tls13.c: In function ‘DoTls13ServerHello’:
src/tls13.c:3611:27: error: ‘tls12minor’ undeclared (first use in this function)
 3611 |         args->pv.minor != tls12minor) {
      |                           ^~~~~~~~~~
```

dgarske commented on src/ssl.c:
```diff
@@ -11930,6 +11991,27 @@ int wolfSSL_DTLS_SetCookieSecret(WOLFSSL* ssl,
+                    && ssl->dtls13Rtx.sendAcks == 1) {
+                    ssl->dtls13Rtx.sendAcks = 0;
+                    /* we aren't negotiated the version yet, so we aren't sure
+                       the other end can speak v1.3. On the other side we have
```
Please use leading `*` for multi-line comment.

dgarske commented on src/ssl.c:
```diff
@@ -11930,6 +11991,27 @@ int wolfSSL_DTLS_SetCookieSecret(WOLFSSL* ssl,
+                       the other end can speak v1.3. On the other side we have
+                       received a unified records, assuming that the ServerHello
+                       got lost, we will send an empty ACK. In case the server
+                       is a DTLS with version less than 1.3, it shoudl just
```
Typo `should`.

dgarske commented on src/internal.c:
```diff
@@ -14845,21 +15184,80 @@ static WC_INLINE word32 UpdateHighwaterMark(word32 cur, word32 first,
 #endif /* WOLFSSL_MULTICAST */
 
+/* diff must be already incremented by one */
+static WC_INLINE void DtlsUpdateWindowGTSeq(word32 diff, word32* window)
```
This is called in two places. Is forcing inline here really appropriate?

rizlik replied:
removed

dgarske commented on src/internal.c:
```diff
@@ -17432,6 +17846,11 @@ int ProcessReplyEx(WOLFSSL* ssl, int allowSocketErr)
         /* get the record layer header */
         case getRecordLayerHeader:
 
+            /* DTLSv1.3 record numbers in the header are encrypted, and AAD
```
Please use leading `*` for multi-line comment.

dgarske commented on src/internal.c:
```diff
@@ -20562,6 +21118,17 @@ int ReceiveData(WOLFSSL* ssl, byte* output, int sz, int peek)
+
+#ifdef WOLFSSL_DTLS13
+        if (ssl->options.dtls) {
+            /* TODO: Dtls13DoScheduledWork(ssl) may return WANT_WRITE, but
```
Is this TODO still valid? Can we fix this TODO?

rizlik replied:
Looking better at it, `ReceiveData` may already return `WANT_WRITE` from other code paths (like in renegotiation). I dropped the TODO and the "not supported" part from the comment.

dgarske commented on src/internal.c:
```diff
@@ -20585,6 +21152,18 @@ int ReceiveData(WOLFSSL* ssl, byte* output, int sz, int peek)
         #endif
+
+#ifdef WOLFSSL_DTLS13
+            /* little wolfSSL_Peek() hack: if wolfSSL_Peek() is invoked with zs
```
Can we avoid documenting as "hack" or improve?

dgarske commented on src/internal.c:
```diff
@@ -20703,12 +21282,32 @@ static int SendAlert_ex(WOLFSSL* ssl, int severity, int type)
+            ret = Dtls13RlAddPlaintextHeader(ssl, output, alert, ALERT_SIZE);
+            if (ret != 0)
+                return ret;
+#endif /* WOLFSSL_DTLS13 */
```
Should there be an `#else` case here? Otherwise this just skips.

rizlik replied:
If we are in the `if` branch, then dtlsv1.3 should be supported, otherwise we should get in the `else` branch. 
Guarding the body of the `if` looks nicer to me but I agree with you that can be dangerous or misleading. I changed it to follow:
```
#ifdef guard
if (condition) {
}
else
#endif 
{
}
```
usual pattern.

dgarske commented on src/dtls13.c:
```diff
@@ -0,0 +1,2428 @@
+ * struct Dtls13HandshakeHeader: represent DTLS Handshake header
+ * @msg_type: type of message (client_hello,server_hello,etc)
+ * @length: length of the message
+ * @messageSeq: message sequence number (used for re-ordering and retranmission)
```
retransmission

dgarske reviewed (changes requested):
Another round of feedback. Very close!

rizlik commented on examples/client/client.c:
```diff
@@ -4029,6 +4119,27 @@ THREAD_RETURN WOLFSSL_THREAD client_test(void* args)
 
+#if defined(HAVE_SESSION_TICKET)
+    while (waitTicket == 1) {
+        unsigned char ticketBuf[256];
```
`SESSION_TICKET_LEN` (`internal.h`) isn't visible from `client.c`. I added:
```
#ifndef SESSION_TICKET_LEN
#define SESSION_TICKET_LEN 256
#endif
```
at the beginning of the file but it's purely aesthetic

julek-wolfssl commented on src/internal.c:
```diff
@@ -20585,6 +21151,17 @@ int ReceiveData(WOLFSSL* ssl, byte* output, int sz, int peek)
         #endif
+
+#ifdef WOLFSSL_DTLS13
+            /* if wolfSSL_Peek() is invoked with zs == 0 it will not block (but
```
"zs" => "sz"

julek-wolfssl reviewed (commented):
Some of the pre-proc macro's are not indented correctly in `server.c` and `wolfio.c` but that is not a blocker.

dgarske reviewed (changes requested):
This is looking perfect. Just one more build combination I'd like to see working.

I think a TLS v1.3 and DTLS v1.3 only build should work.

```
./configure --enable-dtls --enable-dtls13 --disable-tlsv12 && make
src/internal.c:15385:20: error: implicit declaration of function 'DoHandShakeMsgType' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
        if ((ret = DoHandShakeMsgType(ssl, item->msg, &idx, item->type,
                   ^
src/internal.c:15385:20: note: did you mean 'DoTls13HandShakeMsgType'?
./wolfssl/internal.h:1811:20: note: 'DoTls13HandShakeMsgType' declared here
WOLFSSL_LOCAL int  DoTls13HandShakeMsgType(WOLFSSL* ssl, byte* input,
                   ^
src/internal.c:15509:19: error: implicit declaration of function 'DoHandShakeMsgType' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
            ret = DoHandShakeMsgType(ssl, input, inOutIdx, type, size, totalSz);
                  ^
src/internal.c:15603:15: error: implicit declaration of function 'DoHandShakeMsgType' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
        ret = DoHandShakeMsgType(ssl, input, inOutIdx, type, size, totalSz);
              ^
  CC       wolfcrypt/test/testsuite_testsuite_test-test.o
  CC       examples/client/testsuite_testsuite_test-client.o
3 errors generated.
make[2]: *** [src/libwolfssl_la-internal.lo] Error 1
make[2]: *** Waiting for unfinished jobs....
src/tls.c:12518:14: error: No DTLS version enabled!
            #error No DTLS version enabled!
             ^
1 error generated.
```

dgarske commented:
@rizlik a few more issues with g++:

```
./configure CC=g++ --enable-dtls --enable-dtls13 --enable-session-ticket --enable-psk && make

src/tls13.c:1656:9: error: no matching function for call to 'Dtls13HandshakeAddHeader'
        Dtls13HandshakeAddHeader(ssl, output, type, length);
        ^~~~~~~~~~~~~~~~~~~~~~~~
./wolfssl/internal.h:5415:19: note: candidate function not viable: no known conversion from 'byte' (aka 'unsigned char') to 'enum HandShakeType' for 3rd argument
WOLFSSL_LOCAL int Dtls13HandshakeAddHeader(WOLFSSL* ssl, byte* output,
                  ^
src/tls13.c:1683:9: error: no matching function for call to 'Dtls13AddHeaders'
        Dtls13AddHeaders(output, length, type, ssl);
        ^~~~~~~~~~~~~~~~
./wolfssl/internal.h:5396:19: note: candidate function not viable: no known conversion from 'byte' (aka 'unsigned char') to 'enum HandShakeType' for 3rd argument
WOLFSSL_LOCAL int Dtls13AddHeaders(byte* output, word32 length,
                  ^
src/tls13.c:1714:9: error: no matching function for call to 'Dtls13AddHeaders'
        Dtls13AddHeaders(output, length, type, ssl);
        ^~~~~~~~~~~~~~~~
./wolfssl/internal.h:5396:19: note: candidate function not viable: no known conversion from 'byte' (aka 'unsigned char') to 'enum HandShakeType' for 3rd argument
WOLFSSL_LOCAL int Dtls13AddHeaders(byte* output, word32 length,
                  ^

src/dtls13.c:375:9: error: no matching function for call to 'Dtls13MsgWasProcessed'
        Dtls13MsgWasProcessed(ssl, msg->type);
        ^~~~~~~~~~~~~~~~~~~~~
src/dtls13.c:342:13: note: candidate function not viable: no known conversion from 'byte' (aka 'unsigned char') to 'enum HandShakeType' for 2nd argument
static void Dtls13MsgWasProcessed(WOLFSSL* ssl, enum HandShakeType hs)
            ^
src/dtls13.c:817:19: error: no matching function for call to 'Dtls13TypeIsEncrypted'
    isEncrypted = Dtls13TypeIsEncrypted(ssl->dtls13FragHandshakeType);
                  ^~~~~~~~~~~~~~~~~~~~~
src/dtls13.c:229:13: note: candidate function not viable: no known conversion from 'byte' (aka 'unsigned char') to 'enum HandShakeType' for 1st argument
static byte Dtls13TypeIsEncrypted(enum HandShakeType hs_type)
            ^
src/dtls13.c:842:15: error: no matching function for call to 'Dtls13HandshakeAddHeaderFrag'
        ret = Dtls13HandshakeAddHeaderFrag(ssl, output + rlHeaderLength,
              ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
src/dtls13.c:211:12: note: candidate function not viable: no known conversion from 'byte' (aka 'unsigned char') to 'enum HandShakeType' for 3rd argument
static int Dtls13HandshakeAddHeaderFrag(WOLFSSL* ssl, byte* output,
           ^
src/dtls13.c:854:15: error: no matching function for call to 'Dtls13SendOneFragmentRtx'
        ret = Dtls13SendOneFragmentRtx(ssl, ssl->dtls13FragHandshakeType,
              ^~~~~~~~~~~~~~~~~~~~~~~~
src/dtls13.c:778:12: note: candidate function not viable: no known conversion from 'byte' (aka 'unsigned char') to 'enum HandShakeType' for 2nd argument
static int Dtls13SendOneFragmentRtx(WOLFSSL* ssl,
           ^
src/dtls13.c:1323:15: error: no matching function for call to 'Dtls13SendFragment'
        ret = Dtls13SendFragment(ssl, output, sendSz, r->length + headerLength,
              ^~~~~~~~~~~~~~~~~~
src/dtls13.c:461:12: note: candidate function not viable: no known conversion from 'byte' (aka 'unsigned char') to 'enum HandShakeType' for 5th argument
static int Dtls13SendFragment(WOLFSSL* ssl, byte* output, word16 output_size,
           ^
3 warnings generated.
src/dtls13.c:1388:11: error: no matching function for call to 'Dtls13RtxMsgRecvd'
    ret = Dtls13RtxMsgRecvd(ssl, handshake_type, frag_off);
          ^~~~~~~~~~~~~~~~~
src/dtls13.c:719:12: note: candidate function not viable: no known conversion from 'byte' (aka 'unsigned char') to 'enum HandShakeType' for 2nd argument
static int Dtls13RtxMsgRecvd(WOLFSSL* ssl, enum HandShakeType hs,
           ^
src/dtls13.c:1433:5: error: no matching function for call to 'Dtls13MsgWasProcessed'
    Dtls13MsgWasProcessed(ssl, handshake_type);
    ^~~~~~~~~~~~~~~~~~~~~
src/dtls13.c:342:13: note: candidate function not viable: no known conversion from 'byte' (aka 'unsigned char') to 'enum HandShakeType' for 2nd argument
static void Dtls13MsgWasProcessed(WOLFSSL* ssl, enum HandShakeType hs)
            ^
```

These are mostly casting issues where a function expects an enum type but a byte is passed in. Need to add explicit casts in a few places.

dgarske reviewed (approved):
All tests out! Thanks @rizlik , awesome work on this.
* Tested + --enable-all
* Tested + --enable-session-ticket--enable-psk
* Validated a DTLS v1.3/TLS v1.3 only build (--disable-tlsv12)
* Validated use on an embedded target
* Tested using scan-build, CC=g++, valgring, etc...
